### PR TITLE
Rewrite style system

### PR DESCRIPTION
**Pull request by SevenOutman**
## Requirements
- Less >= 3.7 for [`each` function](https://lesscss.org/functions/#list-functions-each) support.

### Button
- Remove `.rs-btn-default` class, merging its styles into `.rs-btn`.

### Misc
- Reduce over-complex compound selectors.
- Remove IE 10 fixes
- Replace `calcFontColor` function mixin with Less `contrast` function
- Rewrite `palette` function mixin as Less plugin. (No need to set `javascriptEnabled: true` for Less anymore)

## BREAKING CHANGES
- Drop `@ns` variable. Customizing global class prefix is not supported now.
- Deprecate `<Navbar.Header>` and `<Navbar.Body>` components.

### CSS custom properties

Most color-related styles are moved from Less variables to CSS custom properties in favor of the new dark mode.

### Dark Mode

In this change, RSuite does not regard dark mode as *a built-in custom theme* anymore. Dark mode is now considered a *color mode* that can be enabled in any custom themes as well as in the default theme. Hence the prebuilt dark CSS is removed and both light and dark CSS files are unified as one.

Before:

- `rsuite-default.css`
- `rsuite-dark.css`

Now:

- `rsuite.css`

### Deprecated/Removed LESS variables

There're some global LESS variables which could be used for customizing styles that have been deprecated or removed in this change. Deprecated variables still take effect but are discouraged for using and will be removed in future updates.

#### Flags are prefixed commonly

There're some boolean variables working as flags for toggling specific features in RSuite styles. The flags are now named with a `@enable-*` prefix.

| Deprecated falg | New flag
| ----- | -----
| `@ie-polyfill` | `@enable-ie-polyfill` 
| `@reset-import` | `@enable-css-reset`
| `@button-ripple` | `@enable-ripple-effect`

#### Stateful colors are removed

- `@success-color`
- `@success-light-color`
- `@info-color`
- `@info-light-color`
- `@warning-color`
- `@warning-light-color`
- `@error-color`
- `@error-light-color`

The variables above were used to define colors for some components with different states, e.g. `<Message>`. Now the state semantics are directed combined with the Secondary Palette, making the color system more consistent.

| State | Corresponding color in Secondary Palette
| ----- | -----
| `success` | `@green` 
| `info` | `@blue`
| `warning` | `@yellow`
| `error` | `@red`

#### Other deprecated variables

| Deprecated variable | Use this instead
| -------------------- | ----------------
| `@base-color` | `@primary-color`
| `@default-box-shadow` | `@box-shadow`
| `@padding-base-vertical` | `@padding-y`
| `@padding-base-horizontal` | `@padding-x`
| `@padding-large-vertical` | `@padding-y-lg`
| `@padding-large-horizontal` | `@padding-x-lg`
| `@padding-small-vertical` | `@padding-y-sm`
| `@padding-small-horizontal` | `@padding-x-sm`
| `@padding-extra-small-vertical` | `@padding-y-xs`
| `@padding-extra-small-horizontal` | `@padding-x-xs`
| `@table-header-content-padding-vertical` | `@table-header-padding-y`
| `@table-header-line-small` | `@table-header-line-height`
| `@table-body-content-padding-vertical` | `@table-cell-padding-y`
| `@table-content-padding-horizontal` | `@table-cell-padding-x`
| `@btn-border-radius-base` | `@btn-border-radius`
| `@btn-ripple-background-color` | `@ripple-bg`
| `@badge-dot-side-length` | `@badge-dot-size`
| `@avatar-md-side-length` | `@avatar-size`
| `@avatar-md-font-size` | `@avatar-font-size`
| `@avatar-lg-side-length` | `@avatar-size-lg`
| `@avatar-lg-font-size` | `@avatar-font-size-lg`
| `@avatar-sm-side-length` | `@avatar-size-sm`
| `@avatar-sm-font-size` | `@avatar-font-size-sm`
| `@avatar-xs-side-length` | `@avatar-size-xs`
| `@avatar-xs-font-size` | `@avatar-font-size-xs`
| `@padding-base-input-vertical` | `@input-padding-y`
| `@padding-base-input-horizontal` | `@input-padding-x`
| `@dropdown-item-padding-vertical` | `@dropdown-item-padding-y`
| `@dropdown-item-padding-horizontal` | `@dropdown-item-padding-x`
| `@nav-bar-padding-vertical` | `@navbar-item-padding-y`
| `@nav-bar-padding-horizontal` | `@navbar-item-padding-x`
| `@nav-bar-height` | `@navbar-height`

### CODE REVIEW

**Comment by hiyangguo** [comment]
@SevenOutman  移除 `@ns` 的时候，应该处理一下 https://github.com/rsuite/rsuite/blob/next/src/utils/useClassNames.ts#L35 的代码吧。感觉 `CustomContext` 中的 `classPrefix` 就不需要了

**Comment by hiyangguo** [comment]
image
@SevenOutman  颜色都没了。我看现在你这边的设计是需要手动添加 `[appearance="primary"]` 或者 `[appearance="subtle"]` 这个 break change 是确定的么？ 我记得当时实现的时候 是需要 默认状态下可以颜色生效的。

**Comment by hiyangguo** [comment]
image
```
.rs-btn {
  .btn-base();

  &:disabled,
  &.rs-btn-disabled {
    &:extend(button[disabled]);
    cursor: @cursor-disabled !important;
    pointer-events: none;

    .rs-ripple-pond {
      display: none !important;
    }
  }
}
```

` pointer-events: none;` 会导致 `cursor` 失效。

**Comment by hiyangguo** [comment]
`<Drawer backdrop="static"/>` 时点击 backdrop 的动画和以前不一样了。很不自然。

**Comment by hiyangguo** [comment]
image
样式异常
link: [http://127.0.0.1:3000/zh/components/drawer/](http://127.0.0.1:3000/zh/components/drawer/)

**Comment by hiyangguo** [comment]
image

**Comment by hiyangguo** [comment]
image

hover 样式异常

**Comment by SevenOutman** [comment]
> image
> @SevenOutman 颜色都没了。我看现在你这边的设计是需要手动添加 `[appearance="primary"]` 或者 `[appearance="subtle"]` 这个 break change 是确定的么？ 我记得当时实现的时候 是需要 默认状态下可以颜色生效的。

这个是确认的，从一开始设计就是只有 default 不支持彩色，其他 appearance 都支持。

**Review by hiyangguo — APPROVED** [review]
> > image
> > @SevenOutman 颜色都没了。我看现在你这边的设计是需要手动添加 `[appearance="primary"]` 或者 `[appearance="subtle"]` 这个 break change 是确定的么？ 我记得当时实现的时候 是需要 默认状态下可以颜色生效的。
> 
> 这个是确认的，从一开始设计就是只有 default 不支持彩色，其他 appearance 都支持，以前的实现是错误的。

那可能是我之前理解错了。

**Comment by hiyangguo** [comment]
@SevenOutman  关于颜色计算这边我当时的想法有两个：
一个是，借助 css calc  ，参考 https://css-tricks.com/css-variables-calc-rgb-enforcing-high-contrast-colors/ 实现  pallete （这个不知道是否可行）
还有一个就是，在 https://rsuitejs.com/tools/palette/ 的生成中进行拓展。选完主色以后，js 计算所有的颜色变量，然后导出。因为纯的 css 变量并不能完全实现 pallatte 的效果。

File: src/Breadcrumb/styles/index.less [hunk 1]
```diff
@@ -1 +1,25 @@
...
+// --------------------------------------------------
+
+.rs-breadcrumb {
+  margin-bottom: @line-height-computed;
+  font-size: @breadcrumb-font-size;
+  color: var(--rs-color-text-secondary);
+
+  // If breadcrumbs has a ci
+  &-item {
+    transition: color 0.3s linear;
```
[the remark this answers is not in the document]
Comment (SevenOutman) [reply src/Breadcrumb/styles/index.less:14]: 我想 `cursor: pointer` 样式应该全局添加在 `button, [role=button], [role=listitem]` 这样的选择器上，而非各个组件上，以提高一致性和可维护性（可能有部分情况除外）。

**Comment by SevenOutman** [comment]
> 一个是，借助 css calc ，参考 https://css-tricks.com/css-variables-calc-rgb-enforcing-high-contrast-colors/ 实现 pallete （这个不知道是否可行）

是个不错的思路。不过，如果使用 `postcss-custom-properties` 来兼容不支持 CSS 变量的浏览器，这里的乘法表达式就会被破坏。

> 还有一个就是，在 https://rsuitejs.com/tools/palette/ 的生成中进行拓展。选完主色以后，js 计算所有的颜色变量，然后导出。因为纯的 css 变量并不能完全实现 pallatte 的效果。

目前是往这个方向考虑，仍通过 JS 来计算。

**Comment by SevenOutman** [comment]
This PR is successfully deployed and ready for review now.

File: src/TagPicker/styles/index.less [hunk 2]
```diff
@@ -1 +1,60 @@
...
+
+  .rs-tag {
+    max-width: ~'calc(100% - @{tag-picker-content-padding-horizontal})';
+    vertical-align: top;
+    .ellipsis-basic();
+  }
+
+  .rs-picker-search-input,
+  .rs-tag {
+    margin-top: @tag-picker-content-padding-vertical;
```
Comment (simonguo) [inline src/TagPicker/styles/index.less:28]: image
tag 的最小号，remove icon 和文本没有对齐

**Comment by simonguo** [comment]
RTL 按钮点击后不能生效，需要刷新页面。 RTL 模式下文档有些布局问题。

File: src/Table/styles/index.less [hunk 3]
```diff
@@ -1 +1,431 @@
...
+    transform: scaleY(1.3);
+  }
+
+  &-vertical {
+    top: 0;
+    right: 0;
+    width: @table-scrollbar-width;
+    // Reserve scroll bar position.
+    bottom: @table-scrollbar-width;
+    background-color: var(--rs-table-scrollbar-vertical-thumb);
```
Comment (simonguo) [inline src/Table/styles/index.less:416]: image
[dark]: scrollbar 与 thumb 的背景颜色一样，导致看不到 thumb

Comment (simonguo) [reply src/Table/styles/index.less:416]: 垂直滚动条

File: src/CheckTreePicker/styles/index.less [hunk 4]
```diff
@@ -1 +1,199 @@
...
+      }
+    }
+  }
+}
+
+.rs-check-tree-node {
+  position: relative;
+  font-size: @picker-tree-node-font-size;
+  line-height: @picker-tree-node-line-height;
+
```
[the remark this answers is not in the document]
Comment (SevenOutman) [reply src/CheckTreePicker/styles/index.less:32]: > 这里需要 `inline-block`

这里是由 `.rs-btn-toolbar` 上的 `line-height: 0` 导致的

[the remark this answers is not in the document]
Comment (SevenOutman) [reply src/CheckTreePicker/styles/index.less:32]: fixed in df013be

File: src/IconButton/styles/index.less [hunk 5]
```diff
@@ -1 +1,54 @@
...
+}
+
+// Only support default/primary button
+.rs-btn-icon-with-text {
+  > .rs-icon {
+    position: absolute;
+    top: 0;
+    display: block;
+    background-color: var(--rs-iconbtn-addon);
+    transition: @btn-transition;
```
[the remark this answers is not in the document]
Comment (SevenOutman) [reply src/IconButton/styles/index.less:27]: fixed in 44430fe

File: src/Table/styles/index.less (hunk 3 above)
Comment (SevenOutman) [reply src/Table/styles/index.less]: fixed in cb50e6d

**Comment by SevenOutman** [comment]
> RTL 模式下文档有些布局问题

是由于编译的 CSS 在进入 `rtlcss` 之前被 `cssnano` 移除了注释，使得一些 `rtlcss:` 指令失效。已修复了。

File: src/Drawer/styles/index.less [hunk 6]
```diff
@@ -1 +1,215 @@
...
+  color: var(--rs-text-heading);
+  font-weight: normal;
+  font-size: @modal-title-font-size;
+  line-height: unit((@line-height-computed / @modal-title-font-size));
+  .ellipsis();
+}
+
+// Footer (for actions)
+.rs-drawer-footer {
+  .clearfix(); // clear it in case folks use .pull-* classes on buttons
```
[the remark this answers is not in the document]
Comment (SevenOutman) [reply src/Drawer/styles/index.less:236]: 重新设计了 Drawer 的标准布局 (486ae72dcb8be3dc70a78d330bff22849a72e97a)

**Pull request merged**